Lib/fontTools/ttLib/tables/otBase.py

```diff
@@ -96,62 +107,95 @@ def compile(self, font):
 						self.tableTag,
 					)
 
+		if (use_hb_repack in (None, True)
+				and have_uharfbuzz
+				and self.tableTag in ("GSUB", "GPOS")):
+			mode = self.MODE_HB_FT
+		else:
+			mode = self.MODE_PURE_FT
+
 		hb_first_error_logged = False
+		lastOverflowRecord = None
 		while True:
 			try:
 				writer = OTTableWriter(tableTag=self.tableTag)
 				self.table.compile(writer, font)
-				if (
-					use_hb_repack in (None, True)
-					and have_uharfbuzz
-					and self.tableTag in ("GSUB", "GPOS")
-				):
-					try:
-						log.debug("serializing '%s' with hb.repack", self.tableTag)
-						return writer.getAllDataUsingHarfbuzz()
-					except (ValueError, MemoryError, hb.RepackerError) as e:
-						# Only log hb repacker errors the first time they occur in
-						# the offset-overflow resolution loop, they are just noisy.
-						# Maybe we can revisit this if/when uharfbuzz actually gives
-						# us more info as to why hb.repack failed...
-						if not hb_first_error_logged:
-							error_msg = f"{type(e).__name__}"
-							if str(e) != "":
-								error_msg += f": {e}"
-							log.warning(
-								"hb.repack failed to serialize '%s', reverting to "
-								"pure-python serializer; the error message was: %s",
-								self.tableTag,
-								error_msg,
-							)
-							hb_first_error_logged = True
-						return writer.getAllData(remove_duplicate=False)
-				return writer.getAllData()
+				if mode == self.MODE_HB_FT:
+					return self.tryPackingHarfbuzz(writer, hb_first_error_logged)
+				elif mode == self.MODE_PURE_FT:
+					return self.tryPackingFontTools(writer)
+				elif mode == self.MODE_FT_FALLBACK:
+					self.tryPackingFontTools(writer)
```

Comment:
at first, it seems strange you are not returning anything when successful (unlike the other two `if` branches above). I assume it's intentional, so it would be good to add a little comment on the expected side effects of `tryPackingFontTools` in the context of `MODE_FT_FALLBACK`

Reply:
Are we sure that `_doneWriting` method can be called multiple times?
In the current code, when `tryPackingHarfbuzz` fails for some hb.repack error, we explictly avoid re-running `_doneWriting` method, because it does `remove_duplicate=False` which skips calling `_doneWriting` => `return writer.getAllData(remove_duplicate=False)`.

`_doneWriting` modifies the OTTableWriter's `self.items` from an unhashable mutable list to a hashable and _immutable_ tuple, so rerunning `_doneWriting` on the same writer instance (no matter the value of `shareExtension` parameter) is either no-op or would incur an error.. Or maybe I'm missing something

/cc @behdad

Reply:
> Are we sure that `_doneWriting` method can be called multiple times? In the current code, when `tryPackingHarfbuzz` fails for some hb.repack error, we explictly avoid re-running `_doneWriting` method, because it does `remove_duplicate=False` which skips calling `_doneWriting` => `return writer.getAllData(remove_duplicate=False)`.
> 
> `_doneWriting` modifies the OTTableWriter's `self.items` from an unhashable mutable list to a hashable and _immutable_ tuple, so rerunning `_doneWriting` on the same writer instance (no matter the value of `shareExtension` parameter) is either no-op or would incur an error.. Or maybe I'm missing something
> 
> /cc @behdad

On every iteration we create and compile a new writer. So when _doneWriting is called again it's safe.